### DISCUSSION TITLE
[Question][Hw3] Handling phi node

### DISCUSSION THREAD

**Discussion by Akais24** [prosyslab-classroom/cs424-program-reasoning#429] [Q&A] [answered]
### Your Name

황재영

### Question

HW3에서 phi node를 핸들링하는 부분에 대해서 이해가 되지 않는 것이 있어서 질문드립니다
README에서 current block에서 파생될 수 있는 다른 block에 phi node가 있으면 세심하게 핸들링하라고 되어 있는데,
README의 내용을 이해한 바에 따르면, 다른 block으로 파생되는 rule에 phi node에서 지정된 값을 expr로 만들어서 넣는다고 생각하였습니다
```
R1: node1(...) /\ I1 /\ ... /\ In => node2(%a)
-> 
R1: node1(...) /\ I1 /\ ... /\ In /\ (%a = 1) => node2(%a)
```

그러나 위와 같이 구현하였다고 생각하였을 때, example4의 케이스에서는 아래와 같은 debug.smt2를 얻게 되는데,
다음 노드를 위해 설정해야 하는 값을, 해당 노드에서 이미 사용하고 있어서 적절하지 않은 규칙이 되고 있는 것처럼 보였습니다
이는 의도한 바가 아닐 것 같아서 제대로 된 solving이 되고 있는 것 같지 않습니다

```
(forall ((%cmp Bool)
         (%call1 Int)
         (%conv Int)
         (%k.0 Int)
         (%call Int)
         (%j.0 Int)
         (%dec Int)
         (%inc Int))
  (=> (and (= %j.0 %inc)
           (= %k.0 %dec)
           (= %dec (+ %k.0 (- 1)))
           (= %inc (+ %j.0 1)))
      (while.cond %dec %inc %j.0 %k.0 %conv %cmp %call1 %call)))
```

이러한 반복적인 변수 사용에 어떻게 대응해야 되는지 궁금합니다

참고. [과거의 유사한 질문](https://github.com/prosyslab-classroom/cs424-program-reasoning/discussions/332)을 찾았으나, 제 상황에 어떻게 적용할지를 몰라서 질문하게 되었습니다

=========================

I have a question about handling phi nodes in HW3 as there's something I don't understand.

The README states that we should carefully handle phi nodes in other blocks that can be derived from the current block. Based on my understanding of the README, I thought that for rules deriving other blocks, we should create an expression using the value specified in the phi node and insert it.
```
R1: node1(...) /\ I1 /\ ... /\ In => node2(%a)
->
R1: node1(...) /\ I1 /\ ... /\ In /\ (%a = 1) => node2(%a)
```
However, when I implemented it this way, for the case in example4, I get the following debug.smt2:

```
(forall ((%cmp Bool)
         (%call1 Int)
         (%conv Int)
         (%k.0 Int)
         (%call Int)
         (%j.0 Int)
         (%dec Int)
         (%inc Int))
  (=> (and (= %j.0 %inc)
           (= %k.0 %dec)
           (= %dec (+ %k.0 (- 1)))
           (= %inc (+ %j.0 1)))
      (while.cond %dec %inc %j.0 %k.0 %conv %cmp %call1 %call)))
```

It seems that the values that should be set for the next node are already being used in that node, resulting in an inappropriate rule. This doesn't seem to be the intended outcome, so I don't think proper solving is happening.
I'm wondering how to handle this repetitive use of variables.

Note: I found a [similar question from the past](https://github.com/prosyslab-classroom/cs424-program-reasoning/discussions/332), but I'm asking because I don't know how to apply it to my situation.

**Comment by spearo2** [comment, marked as answer]
질문하신 부분은 숙제와 직접적인 관련이 있습니다. 이 문제를 스스로 해결하는 것이 숙제의 일부입니다. 

추가적으로, 질문하실때 모든 사람이 질문을 읽고 이해할 수 있도록 영어 번역본을 함께 남겨주세요.

=========================

Your question concerns the important part of the homework. TAs can't help with this type of problem because doing so defeats the purpose. We highly encourage you to find the solution by yourself.

Also, please, leave the English version of your question along so that everyone can read and understand the question.

**Comment by MartinWho2** [comment]
Would it be possible to get a hint on what would be a correct approach to this at least ? I am very confused about how to approach the problem if the variable names are repeated between the two blocks...

**Reply by bongjunj** [reply]
`(x = y) -> S(x)` doesn't mean `S(y)`

**Reply by Akais24** [reply]
For my case, reading previous questions in discussion helped

**Reply by MartinWho2** [reply]
Thanks for the answer ! But now I am confused about something related : How do you choose the order of the parameters for the function call ? :sweat_smile:

**Reply by bongjunj** [reply]
> Thanks for the answer ! But now I am confused about something related : How do you choose the order of the parameters for the function call ? :sweat_smile: 

see how relations are initialized. They are already there :)
